### PR TITLE
 samples: tfm_cert: Add reference TF-M sample for PSA Certified Level 2

### PR DESCRIPTION
**Pull request by joerchan**
Add reference TF-M sample for PSA Certified Level 2.

### CODE REVIEW

File: samples/tfm/tfm_cert/child_image/mcuboot/prj.conf [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+CONFIG_BOOT_ENCRYPT_RSA=n
+CONFIG_BOOT_ENCRYPT_EC256=n
+CONFIG_BOOT_ENCRYPT_X25519=n
+
+CONFIG_BOOT_UPGRADE_ONLY=y # changed from default MCUboot
+CONFIG_BOOT_BOOTSTRAP=n
+CONFIG_MCUBOOT_DOWNGRADE_PREVENTION=y # changed from default MCUBoot
+
+CONFIG_FLASH=y
+CONFIG_FPROTECT=n # changed from default MCUBoot
```
Comment (SebastianBoe) [inline samples/tfm/tfm_cert/child_image/mcuboot/prj.conf:16]: Isn't it desirable for MCUBoot to protect itself and its slots?

File: samples/tfm/tfm_cert/prj.conf [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+CONFIG_MCUMGR_CMD_IMG_MGMT=y
+CONFIG_MCUMGR_CMD_OS_MGMT=y
+CONFIG_MCUMGR_CMD_STAT_MGMT=y
+
+# Enable logging
+CONFIG_LOG=y
+CONFIG_MCUBOOT_UTIL_LOG_LEVEL_WRN=y
+
+# Enable the serial mcumgr transport.
+CONFIG_MCUMGR_SMP_UART=y
```
Comment (SebastianBoe) [inline samples/tfm/tfm_cert/prj.conf:56]: NB: When we are ready, we will need

CONFIG_TFM_NRF_PROVISIONING=y

It's disabled by default

**Comment by SebastianBoe** [comment]
Looking good

File: samples/tfm/tfm_cert/child_image/mcuboot/prj.conf (hunk 1 above)
Comment (joerchan) [reply samples/tfm/tfm_cert/child_image/mcuboot/prj.conf:16]: I got this from hello_world sample, and I didn't investigate why.
Following the fix here: https://github.com/nrfconnect/sdk-nrf/pull/9206 this should now be removed.

File: samples/tfm/tfm_cert/prj.conf [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+CONFIG_SB_SIGNING_KEY_FILE="priv.pem"
+
+# Enable mcumgr.
+CONFIG_MCUMGR=y
+
+# Some command handlers require a large stack.
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2304
+CONFIG_MAIN_STACK_SIZE=2048
+
+# Ensure an MCUboot-compatible binary is generated.
```
Comment (mswarowsky) [inline samples/tfm/tfm_cert/prj.conf:30]:
```suggestion
# Enable MCUboot so the application is built as MCUboot-compatible image
```

File: samples/tfm/tfm_cert/prj.conf [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+CONFIG_TFM_LOG_LEVEL_SILENCE=n
+CONFIG_TFM_SPM_LOG_LEVEL_DEBUG=y
+CONFIG_TFM_EXCEPTION_INFO_DUMP=y
+
+CONFIG_SECURE_BOOT=y
+CONFIG_BUILD_S1_VARIANT=n
+
+CONFIG_FW_INFO=y
+CONFIG_FW_INFO_FIRMWARE_VERSION=1
```
Comment (mswarowsky) [inline samples/tfm/tfm_cert/prj.conf:18]: Maybe I'm wrong but this should set the version for NISB, so it should be set for the MCUboot image, as the MCUboo image is the one that gets booted and "upgraded" by NISB. 
And for the application we should set the MCUBoot_version

File: samples/tfm/tfm_cert/sample.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+common:
+    tags: tfm
+    platform_allow: nrf5340dk_nrf5340_cpuapp_ns nrf9160dk_nrf9160_ns
+    integration_platforms:
+      - nrf5340dk_nrf5340_cpuapp_ns
+      - nrf9160dk_nrf9160_ns
+    harness: console
+    harness_config:
+      type: multi_line
+      regex:
```
Comment (mswarowsky) [inline samples/tfm/tfm_cert/sample.yaml:13]: Here we should a some lines of the output, like:
`Requesting initial attestation token with 64 byte challenge`
`Received initial attestation token of 360 bytes.`

**Review by mswarowsky — CHANGES_REQUESTED** [review]

File: samples/tfm/tfm_cert/src/main.c [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+#ifdef CONFIG_MCUMGR_CMD_SHELL_MGMT
+#include "shell_mgmt/shell_mgmt.h"
+#endif
+#ifdef CONFIG_MCUMGR_CMD_FS_MGMT
+#include "fs_mgmt/fs_mgmt.h"
+#endif
+
+#define LOG_LEVEL LOG_LEVEL_DBG
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(smp_sample);
```
Comment (Vge0rge) [inline samples/tfm/tfm_cert/src/main.c:40]: Can we not include the log level as a second argument of LOG_MODULE_REGISTER(... as we do in the other samples

File: samples/tfm/tfm_cert/src/main.c [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+static void get_fw_info_address(uint32_t fw_address)
+{
+	struct fw_info info;
+	int err;
+
+	err = tfm_platform_firmware_info(fw_address, &info);
+	if (err) {
+		printk("Failed to retrieve fw_info for address %u\n", fw_address);
```
Comment (Vge0rge) [inline samples/tfm/tfm_cert/src/main.c:109]: Since we have zephyr logging here why are we using printk?

File: samples/tfm/tfm_cert/src/main.c [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+	printk("Magic: 0x");
+	for (int i = 0; i < MAGIC_LEN_WORDS; i++) {
+		printk("%08x", info.magic[i]);
+	}
+	printk("\n");
+
+	printk("Total Size: %d\n", info.total_size);
+	printk("Size: 0x%08x\n", info.size);
+	printk("Version: %u\n", info.version);
+	printk("address: 0x%08x\n", info.address);
```
Comment (Vge0rge) [inline samples/tfm/tfm_cert/src/main.c:122]: Nit: Capitalize letters

File: samples/tfm/tfm_cert/README.rst [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    -DCONFIG_BOOT_SIGNATURE_KEY_FILE="/home/user/ncs/_keys/nsib_priv.pem"
+    -Dmcuboot_CONFIG_BOOT_SIGNATURE_KEY_FILE=\"/home/user/ncs/_keys/mcuboot_priv.pem\" \
+
+See :ref:`ug_fw_update_keys` for information on how to generate custom keys for a project.
+
+The bootloader and application can be updated using the The :file:`mcumgr` command-line tool.
+See :ref:`smp_svr_sample` for installation instructions and usage instructions.
+
+Application and TF-M Firmware upgrade
+*************************************
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:63]:
```suggestion
Application and TF-M firmware update
************************************
```

File: samples/tfm/tfm_cert/README.rst [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+The application and TF-M firmware can be updated using firmware upgrade.
+For the image to be updatable the firmware image version :kconfig:option:`CONFIG_MCUBOOT_IMAGE_VERSION` has to be updated to a higher version.
+See :ref:`ug_fw_update_image_versions_mcuboot_downgrade` for information on downgrade protection in MCUBoot.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:67]:
```suggestion
Use firmware update to update the application and TF-M firmware.
For the image to be updatable, the firmware image version :kconfig:option:`CONFIG_MCUBOOT_IMAGE_VERSION` has to be updated to a higher version.
See :ref:`ug_fw_update_image_versions_mcuboot_downgrade` for information on downgrade protection in MCUboot.
```

File: samples/tfm/tfm_cert/README.rst [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+To upload a new application image we need to build an application with an updated image version.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:69]:
```suggestion
To upload a new application image, build an application with an updated image version.
```

File: samples/tfm/tfm_cert/README.rst [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+
+Then upload the new application image to the device.
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image list
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image upload build_update/zephyr/app_update.bin
+
+Once the new application image is uploaded the hash of the image will be shown in the image listing.
+Flag the image to be tested on next reboot using its hash.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:84]:
```suggestion
Once the new application image is uploaded, the hash of the image is shown in the image list.
Flag the image to be tested on next reboot using its hash.
```

File: samples/tfm/tfm_cert/README.rst [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image list
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image test <hash>
+
+We can then trigger the application upgrade by initiating a reset, the verification of the image will happen during the upgrade process.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:91]:
```suggestion
Trigger the application update by initiating a reset.
The verification of the image will happen during the update process.
```

File: samples/tfm/tfm_cert/README.rst [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 reset
+
+Bootloader Firmware upgrade
+***************************
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:98]:
```suggestion
Bootloader firmware update
**************************
```

File: samples/tfm/tfm_cert/README.rst [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+
+To upload a new bootloader image we need to build a bootloader targeting the correct bootloader slot with an updated firmware image version.
+By default the bootloader is placed in slot 0, so build enable building of the slot 1 bootloader.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:101]:
```suggestion
To upload a new bootloader image, build a bootloader targeting the correct bootloader slot with an updated firmware image version.
The bootloader is placed in slot 0 by default, so enable building of the slot 1 bootloader.
```

File: samples/tfm/tfm_cert/README.rst [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+
+    west build -b nrf5340dk_nrf5340_cpuapp_ns samples/tfm/tfm_cert \
+    -DCONFIG_BUILD_S1_VARIANT=y \
+    -DFW_INFO_FIRMWARE_VERSION=2
+
+List the current firmware images and upload a bootloader image that targets thet non-active bootloader slot.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:107]:
```suggestion
List the current firmware images and upload a bootloader image that targets the non-active bootloader slot.
```

File: samples/tfm/tfm_cert/README.rst [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image list
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image upload build/zephyr/signed_by_mcuboot_and_b0_s1_image_update.bin
+
+Once the new bootloader image is uploaded the hash of the image will be shown in the image listing.
+Flag the image to be tested on next reboot using its hash.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:115]:
```suggestion
Once the new bootloader image is uploaded, the hash of the image is shown in the image list.
Flag the image to be tested on next reboot using its hash.
```

File: samples/tfm/tfm_cert/README.rst [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image list
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 image test <hash>
+
+We can then trigger the bootloader upgrade by initiating a reset, the verification of the image will happen during the upgrade process.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:122]:
```suggestion
Trigger the bootloader update by initiating a reset.
The verification of the image will happen during the update process.
```

File: samples/tfm/tfm_cert/README.rst [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+    west build -b nrf5340dk_nrf5340_cpuapp samples/tfm/provisioning_image -d build_provisioning_image
+    west flash --erase -d build_provisioning_image
+
+Build and flash the TF-M PSA Certified Level 2 sample, do not flash with ``--erase`` or the PSA platform security parameters will be lost.
+
+.. code-block:: console
+
+    west build -b nrf5340dk_nrf5340_cpuapp_ns samples/tfm/tfm_cert
+    west flash
```
[the remark this answers is not in the document]
Comment (mswarowsky) [reply samples/tfm/tfm_cert/README.rst:42]: Ah got dam that comment should be in the line of building the provisioning sample

File: samples/tfm/tfm_cert/README.rst [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+Overview
+********
+
+This sample uses Trusted Firmware-M, nRF Secure Immutable bootloader and MCUboot to satisfy the Platform Security Architecture (PSA) Certified Level 2 requirements.
+The sample prints information about the identity of the device and the firmware versions that are currently running.
+
+Building and running
+********************
+
+The sample requires that the device is provisioned using the :ref:`provisioning image <provisioning_image>` sample.
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:28]: When are we counting on having merged [provisioning image sample ](https://github.com/nrfconnect/sdk-nrf/pull/9033)? Doc build for this PR will keep failing due to the link until the provisioning image sample is merged into main.

File: samples/tfm/tfm_cert/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+project(tfm_cert)
+
+# NORDIC SDK APP START
+target_sources(app PRIVATE src/main.c)
+# NORDIC SDK APP END
+
```
[the remark this answers is not in the document]
Comment (joerchan) [reply samples/tfm/tfm_cert/CMakeLists.txt:16]: default keys of the sample is removed.
The build system has warning for mcuboot default keys.
Added PR for nsib generated keys: https://github.com/nrfconnect/sdk-nrf/pull/9417

File: samples/tfm/tfm_cert/README.rst [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+    west build -b nrf5340dk_nrf5340_cpuapp_ns samples/tfm/tfm_cert
+    west flash
+
+Firmware Upgrade
+****************
+
+The sample supports firmware upgrade of both the application and TF-M, and the second stage bootloader.
+The firmware update process requires signature verification keys in order to sign the images used in the firmware update process.
+The sample is supplied with its own set of private keys for signing.
+These keys can be replaced with custom keys.
```
[the remark this answers is not in the document]
Comment (joerchan) [reply samples/tfm/tfm_cert/README.rst]: Removed, and also mentioned that MCUBoot and NSIB use signing keys by default that should not be used for production.

File: samples/tfm/tfm_cert/child_image/mcuboot/prj.conf [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+CONFIG_LOG=y
+CONFIG_LOG_MODE_MINIMAL=y
+### Ensure Zephyr logging changes don't use more resources
+CONFIG_LOG_DEFAULT_LEVEL=0
+### Decrease footprint by ~4 KB in comparison to CBPRINTF_COMPLETE=y
+CONFIG_CBPRINTF_NANO=y
+CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT=0
+
+# python3 bootloader/mcuboot/scripts/imgtool.py keygen -t rsa-2048 -k priv.pem
+CONFIG_BOOT_SIGNATURE_TYPE_RSA=y
```
[the remark this answers is not in the document]
Comment (joerchan) [reply samples/tfm/tfm_cert/child_image/mcuboot/prj.conf]: Updated to ECDSA for this sample.

File: samples/tfm/tfm_cert/prj.conf [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+# python3 nrf/scripts/bootloader/keygen.py --private -o priv.pem
```
[the remark this answers is not in the document]
Comment (joerchan) [reply samples/tfm/tfm_cert/prj.conf]: Moved to readme.

File: samples/tfm/tfm_cert/README.rst (hunk 20 above)
Comment (joerchan) [reply samples/tfm/tfm_cert/README.rst:28]: Fingers crossed, very soon.

File: samples/tfm/tfm_cert/README.rst [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+
+    CONFIG_SB_SIGNING_KEY_FILE="/home/user/ncs/_keys/nsib_priv.pem"
+
+See :ref:`ug_fw_update_keys` for more information on how to generate and use keys for a project.
+
+The bootloader and the application can be updated using the :file:`mcumgr` command-line tool.
+See :ref:`smp_svr_sample` for installation and usage instructions.
+
+Application and TF-M firmware update
+************************************
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:82]: This should probably be a subsection, right?
```suggestion
====================================
```

File: samples/tfm/tfm_cert/README.rst [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+
+Trigger the application update by initiating a reset.
+The verification of the image will happen during the update process.
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 reset
+
+Bootloader firmware update
+**************************
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:118]: Subsection?
```suggestion
==========================
```

File: samples/tfm/tfm_cert/README.rst [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+
+Trigger the bootloader update by initiating a reset.
+The verification of the image will happen during the update process.
+
+.. code-block:: console
+
+    mcumgr --conntype serial --connstring dev=/dev/ttyACM2,baud=115200,mtu=512 reset
+
+Testing
+=======
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:150]: This section is usually a subsection of Building and running. Maybe it should be moved to before the Firmware update section.

File: samples/tfm/tfm_cert/README.rst [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+
+This sample is a reference for how to achieve PSA Certified Level 2 using TF-M on nRF devices.
+
+Requirements
+************
+
+The sample supports the following development kits:
+
+.. table-from-sample-yaml::
+
```
Comment (mia-ko) [inline samples/tfm/tfm_cert/README.rst:18]: Do we need .. include:: /includes/tfm.txt here?

**Review by mia-ko — CHANGES_REQUESTED** [review]

**Review by mswarowsky — APPROVED** [review]
Only a comment change left , the rest looks fine for me

File: samples/tfm/tfm_cert/sample.yaml (hunk 5 above)
Comment (joerchan) [reply samples/tfm/tfm_cert/sample.yaml:13]: Created follow-up task for automated testing of this sample.

**Review by mia-ko — COMMENTED** [review]
Missing entry in release notes for this sample.
Doc build will be failing due to link to a sample that is not yet merged into main.
Otherwise looks ok :)

**Review by mia-ko — APPROVED** [review]

**Pull request merged**